Cabal/src/Distribution/PackageDescription/Check.hs

```diff
@@ -2120,7 +2135,8 @@ data CheckPackageContentOps m = CheckPackageContentOps {
     doesFileExist        :: FilePath -> m Bool,
     doesDirectoryExist   :: FilePath -> m Bool,
     getDirectoryContents :: FilePath -> m [FilePath],
-    getFileContents      :: FilePath -> m BS.ByteString
+    getFileContents      :: FilePath -> m BS.ByteString,
+    getGlobFiles         :: FilePath -> FilePath -> m [FilePath]
```

Comment:
Does this have to be added to the interface? This has to be replicated for `.tar.gz` files somehow (at least), at server level. Will there be a sensible/easy implementation for that?

Reply:
Ok, I am sorry this is definitely *not* working. Let me summarise my finding:
- take your test folder and `cabal sdist` it
- then take the `tar.gz` file and `tar xf` it
- you will get a folder with a changelog file but *not* a readme one.

What I mean is that the check will **not** work on a `tar.gz`, which is what CheckPackageContentOps abstracts.

Heed the comments on `checkPackageFilesPreDistribution`.

Reply:
> Does this have to be added to the interface?

We do not have direct access to `IO` in `checkPackageContent` and I need `runDirFileGlob`, so I would say yes. However I do not have a strong opinion on whether we should put this check in `checkPackageFilesPreDistribution` rather than in `checkPackageContent`, so if you think the former is more appropriate I will follow your advice.

> * you will get a folder with a changelog file but _not_ a readme one.

It _is_ expected here: this test warns about `README` existing but not included in the cabal file, while the changelog is.

Reply:
One question though: if using `checkPackageFilesPreDistribution`, will I get a warning _on Hackage_ when I upload a package candidate? Because if this check is only done manually using `cabal check`, then I doubt it will be useful.

Reply:
> It is expected here: this test warns about README existing but not included in the cabal file, while the changelog is.

What I mean is that *there will be no README file in the `.tar.gz`*. Without a README present in the `.tar.gz`, there is no way to *warn it is missing from `extra-doc-files`*.

This is the crucial point, let me know whether I should elaborate it further.

>  One question though: if using checkPackageFilesPreDistribution, will I get a warning on Hackage when I upload a package candidate?

You wouldn’t get a warning in either case upon uploading a candidate to hackage, as the server is not aware of a — e.g. — missing README.

Reply:
I get it now, thanks.

Reply:
I updated the code to address your comment.

Reply:
I think @Bodigrim’s [comment](https://github.com/haskell/cabal/pull/8657#issuecomment-1377999851) is very relevant: in this case we should report either a file omitted in globs (as now) or an inexistent file (new).

Reply:
If you take that approach (missing **or** omitted readme), you can basically do it *without* checking the working tree. Is xyz missing from `extra-doc-files`? Output warning.

Reply:
Yes. I fear this new warning will break lots of tests though.

Reply:
That is sad and true, it will break a bazillion tests.

Reply:
Removed `getGlobFiles` from the interface.